pkg/build/controller/common/util.go

```diff
@@ -42,6 +48,112 @@ func HandleBuildCompletion(build *buildapi.Build, runPolicies []policy.RunPolicy
 	if err := runPolicy.OnComplete(build); err != nil {
 		utilruntime.HandleError(fmt.Errorf("failed to run policy on completed build %s/%s: %v", build.Namespace, build.Name, err))
 	}
+	if err := HandleBuildPruning(build, buildLister, buildConfigLister, buildDeleter); err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to prune old builds %s/%s: %v", build.Namespace, build.Name, err))
+	}
+}
+
+// HandleBuildPruning handles the deletion of old successful and failed builds
+// based on settings in the BuildConfig.
+func HandleBuildPruning(build *buildapi.Build, buildLister buildclient.BuildLister, buildConfigLister buildclient.BuildConfigLister, buildDeleter pruner.BuildDeleter) error {
+	if build.Status.Config == nil {
```

Comment:
I think you should do the following: use util.ConfigNameForBuild to get the bc name from the build. Then use a .Get() to get the single bc without listing.